pkg/R/RDD.R

```diff
@@ -1039,6 +1039,60 @@ setMethod("keyBy",
             lapply(rdd, apply.func)
           })
 
+#' Return a new RDD that has exactly numPartitions partitions.
+#' Can increase or decrease the level of parallelism in this RDD. Internally,
+#' this uses a shuffle to redistribute data.
+#' If you are decreasing the number of partitions in this RDD, consider using
+#' coalesce, which can avoid performing a shuffle.
+#'
+#' @param rdd The RDD.
+#' @param numPartitions Number of partitions to create.
+#' @rdname repartition
+#' @seealso coalesce
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' rdd <- parallelize(sc, list(1, 2, 3, 4, 5, 6, 7), 4L)
+#' numPartitions(rdd)                   # 4
+#' numPartitions(repartition(rdd, 2L))  # 2
+#'}
+setGeneric("repartition", function(rdd, numPartitions) { standardGeneric("repartition") })
+
+#' @rdname repartition
+#' @aliases repartition,RDD
+setMethod("repartition",
+          signature(rdd = "RDD", numPartitions = "integer"),
+          function(rdd, numPartitions) {
+            jrdd <- callJMethod(getJRDD(rdd), "repartition", numPartitions)
```

Comment:
I'm not sure this works as required. This is mainly because in SparkR if the data is encoded as `byte` we only have 1 record in each partition. (The entire partition is encoded as a single byte array). So when the JVM tries to hash keys to partitions we will end up with entire partitions moving together. 

I think the way to get around this is to create a PairRDD just like its created in Scala [1] and then call `partitionBy` and pass in the number of partitions.
 
[1] https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/RDD.scala#L358